fire service protocol extension events for frames

devoncarew opened the pull request:
We're currently recording frame info to the VM timeline; this PR also fires service protocol extension events for flutter frames. The event includes the `frameNumber`, the `startTime` (in micros, from app start), and `elapsed` time, in micros. Debugger clients can opt-into this event via `streamListen('Extension')`.

This info will allow clients to show a bit more awareness of real-time frame times.

@Hixie

devoncarew committed:
fire service protocol extension events for frames

devoncarew committed:
start time in micros

Hixie commented on packages/flutter/lib/src/scheduler/binding.dart:
```diff
@@ -551,6 +551,7 @@ abstract class SchedulerBinding extends BindingBase {
   Duration _currentFrameTimeStamp;
 
   int _debugFrameNumber = 0;
+  Stopwatch _frameStopwatch;
```
if this is debug-only, please name it _debugFrameStopwatch

Hixie commented:
Seems a bit dubious to be recording this information in debug mode, since we don't make particular efforts to get good performance in debug builds. Ideally this would be profile-mode-only.

devoncarew commented:
> Ideally this would be profile-mode-only.

Ah, yeah, we would want the events when in profile mode. Is there a way to recognize that in the framework? (I'm realizing that the asserts will all be stripped out at that point)

Hixie commented:
I think that's a Dart-level problem. Ideally I'd like a way to have profile-mode-only code stripped out of the binary in release builds.

devoncarew commented:
We have a way today to recognize debug mode (via `assert()`s). I'm guessing we don't have a way to differentiate between profile and release modes? Ideally I think I'd want the frame events available in debug and profile modes (profile for accurate, realtime frame info, and debug mode because the info is still interesting, and it would allow tools to give a more live sense of the running app.

Hixie commented:
I suggest we create a function in the foundation layer called `profile` that takes a `VoidCallback` and that only calls its argument if we're in profile mode. Then, we should file a bug on the VM asking for a way to guarantee (with a test) that this function is only included in profile mode, and that it is entirely omitted (including even the jump into the function) in release builds.

devoncarew commented:
sgtm; I'm in the process of re-working this given the `const bool.fromEnvironment("dart.vm.product")` test. An initial version of `profile()` could be based on that.

devoncarew committed:
Merge branch 'master' into fire_events_for_frames

devoncarew committed:
introduce a profile() function; only send frame events when in profile (or debug) modes

devoncarew commented:
@Hixie, re-worked, but I could use some guidance about where `profile()` should live.

devoncarew committed:
moved the profile() function to foundation/profile.dart

devoncarew commented:
(I added `profile()` to `foundation/profile.dart`)

Hixie commented on packages/flutter/lib/src/foundation/profile.dart:
```diff
@@ -0,0 +1,18 @@
+/// When running in profile mode (or debug mode), invoke the given function.
+///
+/// In release mode, the function is not invoked. In the future, we'd want the
+/// given closure - and the call to [profile] - to be tree-shaken out.
```
The "In the future" sentence should probably just be a TODO() rather than part of the documentation.

Hixie commented:
This looks fine but needs a test for the Flutter.Frame logic (verifying that the data is being sent in the expected format, for example, so that nobody accidentally breaks the protocol during a global search-and-replace of the string "startTime" or some such).

devoncarew commented:
> This looks fine but needs a test for the Flutter.Frame logic

👍, will add a test

devoncarew commented:
Just for context, here is the real-time display I'm playing with in IntelliJ:

<img width="296" alt="screen shot 2017-07-18 at 6 57 41 pm" src="https://user-images.githubusercontent.com/1269969/28347287-0e751d6a-6beb-11e7-9e2a-1188c2b0279f.png">

devoncarew committed:
Merge branch 'master' into fire_events_for_frames

devoncarew committed:
refactor to make the change more testable; test the change

devoncarew committed:
Merge branch 'master' into fire_events_for_frames

sethladd commented:
cc @dantup for ideas for VS Code :)

@devoncarew this looks great, thanks!

devoncarew commented:
@Hixie, I did refactor this a bit for better testability; we now have an event stream for the frame information, and that gets copied into service protocol events. It's not exactly the test you mentioned above, but does test that we're creating frame info events w/ the expected info.

```
Stream<FrameInfo> get onFrameInfo;
```

Hixie commented on packages/flutter/lib/src/scheduler/binding.dart:
```diff
@@ -716,3 +744,18 @@ bool defaultSchedulingStrategy({ int priority, SchedulerBinding scheduler }) {
   return true;
 }
+
+/// Performance information for a Flutter frame.
```
we need much more information here, such as, how to get an instance of such a class, what information it stores, what modes it would be available in, the protocol by which it's sent over to IDEs or pointers to documentation on that protocol, etc.

Hixie commented on packages/flutter/lib/src/scheduler/binding.dart:
```diff
@@ -716,3 +744,18 @@ bool defaultSchedulingStrategy({ int priority, SchedulerBinding scheduler }) {
+
+  /// The frame number.
+  final int number;
+  /// The start time in microseconds.
```
blank lines between members

Hixie replied:
"start time in microseconds" only raises more questions! relative to what? is it monotonic? is it affected by timeDilation? etc etc etc

Hixie commented on packages/flutter/lib/src/scheduler/binding.dart:
```diff
@@ -716,3 +744,18 @@ bool defaultSchedulingStrategy({ int priority, SchedulerBinding scheduler }) {
+class FrameInfo {
+  FrameInfo(this.number, this.startTime, this.elapsed);
+
+  /// The frame number.
```
this documentation doesn't comply to our style guide.

Hixie commented on packages/flutter/lib/src/scheduler/binding.dart:
```diff
@@ -716,3 +744,18 @@ bool defaultSchedulingStrategy({ int priority, SchedulerBinding scheduler }) {
+  final int elapsed;
+
+  @override
+  String toString() => 'frame $number, elapsed time: $elapsed micros';
```
rather than "micros" let's just say "µs".

Hixie commented on packages/flutter/lib/src/scheduler/binding.dart:
```diff
@@ -550,9 +550,18 @@ abstract class SchedulerBinding extends BindingBase {
-  int _debugFrameNumber = 0;
+  int _profileFrameNumber = 0;
+  Stopwatch _profileFrameStopwatch = new Stopwatch();
+  StreamController<FrameInfo> _frameInfoController = new StreamController<FrameInfo>.broadcast();
```
let's not initialise this in release mode...

Hixie replied:
also presumably it should be final?

Hixie commented on packages/flutter/lib/src/scheduler/binding.dart:
```diff
@@ -550,9 +550,18 @@ abstract class SchedulerBinding extends BindingBase {
+  /// These events are created after a frame is rendered and contain performance related
+  /// information. Events are only fired when the app is running in debug or profile mode;
+  /// no frame information events are fired in release mode.
+  Stream<FrameInfo> get onFrameInfo => _frameInfoController.stream;
```
this member doesn't comply with our style guide's naming convention.

Hixie commented:
i'm not entirely convinced that exposing this stream to the app makes much sense... what's the use case other than tests?

devoncarew committed:
fire service protocol events by listening to onFrameInfo

devoncarew commented:
> i'm not entirely convinced that exposing this stream to the app makes much sense... what's the use case other than tests?

I don't have any other use case in mind than for tests. It looks like we do expose some other timing related information in `SchedulerBinding`. I'm also happy to try and reduce the visibility of the API to get at the stream of event info.

Hixie commented:
I'm just dubious about instantiating a whole Stream object and so forth in applications, just for testing purposes. That's a very high cost for something that happens very rarely.

Hixie commented:
Maybe you could collect the data then pass it to a protected method that calls developer.whatever, and in the test just subclass the binding and override that method?

The real thing that needs testing though is the VM service protocol and none of these tests are going to do that...

devoncarew commented:
> Maybe you could collect the data then pass it to a protected method that calls developer.whatever, and in the test just subclass the binding and override that method?

I'll do some noodling, and either rework for visibility, or move to another type of test (a devicelab test) which tests at a different level.

> The real thing that needs testing though is the VM service protocol and none of these tests are going to do that...

I agree it would be good to test at that level - not sure that we currently are for our other service extension work.

Hixie commented:
> not sure that we currently are for our other service extension work.

Someone has to be the first. :-)

For example, until yesterday, we didn't have any test of `flutter run` that actually tested that pressing keys did anything. So when testing my new "show performance overlay" feature, rather than testing it the way we test the other features, I created a new kind of test that runs `flutter run`, and writes to its `stdin` stream, then runs `flutter drive` tests to see if it's having any effect. Bam, new kind of test. :-)

DanTup commented:
@sethladd I don't believe Code has any way for us to render arbitrary content like this :(

Also; we don't currently support profiling mode (I don't really know much about this) which I figure is where this makes sense. I've raised https://github.com/Dart-Code/Dart-Code/issues/373 to have this in the list though :-)

devoncarew committed:
Merge branch 'master' into fire_events_for_frames

devoncarew committed:
remove the frame event stream; add a devicelab test

devoncarew committed:
remove a todo

devoncarew committed:
final

devoncarew commented:
@Hixie - circling back to this one - I updated it to add a devicelab test, and removed the public API around listening for frame events (we're back to only firing them on the service protocol).

Hixie commented on dev/devicelab/bin/tasks/service_extensions_test.dart:
```diff
@@ -0,0 +1,88 @@
+      expect(event.data['startTime'] is int);
+      expect(event.data['startTime'] >= 0);
+      expect(event.data['elapsed'] is int);
+      expect(event.data['elapsed'] >= 0);
```
this is fantastic. thank you!

Hixie commented:
This is wonderful.
![LGTM](http://junkyard.damowmow.com/577)

devoncarew commented:
bombs away

devoncarew commented:
Looks like this is turning the devicelab bots red (https://flutter-dashboard.appspot.com/api/get-log?ownerKey=ahNzfmZsdXR0ZXItZGFzaGJvYXJkclgLEglDaGVja2xpc3QiOGZsdXR0ZXIvZmx1dHRlci80YjRjYWJiNzYxZWMxY2I3ZmI1ZjJlOGVjNmU4NTBkZTI5YTAzMDU2DAsSBFRhc2sYgICAgICAyAsM). I'm not sure why, so may revert in the meantime,
